src/py.rs

```diff
@@ -0,0 +1,337 @@
+        let v_hash = k.as_ref(py).hash()?;
+        self.map.entry(k_hash).or_default().insert(v_hash, (k, v));
+        Ok(())
+    }
+
+    /// Gets item by key
+    pub fn get_item<K: ToBorrowedObject>(&self, py: Python, k: K) -> PyResult<Option<PyObject>> {
+        let k = as_dict_obj(py, k.to_object(py))?;
+        let k_hash = k.as_ref(py).hash()?;
+        let bucket = match self.map.get(&k_hash) {
+            Some(bucket) => bucket,
+            None => return Ok(None),
+        };
+        Ok(bucket
+            .values()
+            .find(|(bucket_k, _)| Self::is_eq(py, &k, bucket_k))
+            .map(|(_, v)| v.clone()))
+    }
+
+    /// Remove item by key
+    pub fn remove_item<K: ToBorrowedObject>(
+        &mut self,
+        py: Python,
+        k: K,
+    ) -> PyResult<Option<PyObject>> {
+        let k = as_dict_obj(py, k.to_object(py))?;
+        let k_hash = k.as_ref(py).hash()?;
+        let bucket = match self.map.get_mut(&k_hash) {
+            Some(bucket) => bucket,
+            None => return Ok(None),
+        };
+        Ok(bucket
+            .values_mut()
+            .find(|(bucket_k, _)| Self::is_eq(py, &k, bucket_k))
+            .map(|(_, v)| v.clone()))
+    }
+
+    /// Iterator over both keys and values
+    pub fn iter<'py>(
+        &'py self,
+        py: Python<'py>,
+    ) -> impl Iterator<Item = (&'py PyAny, &'py PyAny)> + 'py {
+        self.map
+            .values()
+            .flat_map(BTreeMap::values)
+            .map(move |(k, v)| (k.clone().into_ref(py), v.clone().into_ref(py)))
+    }
+
+    /// Iterator which consumes object
+    pub fn into_iter(self, py: Python<'_>) -> impl Iterator<Item = (&'_ PyAny, &'_ PyAny)> {
+        self.map
+            .into_iter()
+            .flat_map(|(_, bucket)| bucket.into_iter())
+            .map(move |(_, (k, v))| (k.into_ref(py), v.into_ref(py)))
+    }
+
+    /// Iterator over keys
+    pub fn keys<'py>(&'py self, py: Python<'py>) -> impl Iterator<Item = &'py PyAny> + 'py {
+        self.iter(py).map(|(k, _)| k)
+    }
+
+    /// Iterator over keys which consumes object
+    pub fn into_keys(self, py: Python<'_>) -> impl Iterator<Item = &'_ PyAny> {
+        self.into_iter(py).map(|(k, _)| k)
+    }
+
+    /// Iterator over values
+    pub fn values<'py>(&'py self, py: Python<'py>) -> impl Iterator<Item = &'py PyAny> + 'py {
+        self.iter(py).map(|(_, v)| v)
+    }
+
+    /// Iterator over values which consumes object
+    pub fn into_values(self, py: Python<'_>) -> impl Iterator<Item = &'_ PyAny> {
+        self.into_iter(py).map(|(_, v)| v)
+    }
+
+    /// Hashes object
+    pub fn hash(&self, py: Python) -> Hash {
+        let sum_hash =
+            |a: isize, b: isize| a.wrapping_add(b).into_py(py).as_ref(py).hash().unwrap();
+
+        self.map
+            .iter()
+            .flat_map(|(key, bucket)| bucket.keys().copied().zip(iter::repeat(*key)))
+            .fold(0, |prev, (value, key)| sum_hash(prev, sum_hash(value, key)))
```

Comment:
Maybe you can put here a link to the generalized this hash function description and why it is good?

Reply:
Rewrote it to make it easier to read